src/exec/mod.rs

```diff
@@ -0,0 +1,317 @@
+                    }
+                }
+                ArgOrFieldCode::IconKey => self.icon().map(ToString::to_string),
+                ArgOrFieldCode::TranslatedName => {
+                    let locale = std::env::var("LANG").ok();
+                    if let Some(locale) = locale {
+                        let locale = locale.split_once('.').map(|(locale, _)| locale);
+                        self.name(locale).map(|locale| locale.to_string())
+                    } else {
+                        None
+                    }
+                }
+                ArgOrFieldCode::DesktopFileLocation => {
+                    Some(self.path.to_string_lossy().to_string())
+                }
+                ArgOrFieldCode::Arg(arg) => Some(arg.to_string()),
+            })
+            .collect()
+    }
+}
+
+fn with_non_default_gpu(mut cmd: Command) -> Command {
+    let gpus = Gpus::load();
+    let gpu = if gpus.is_switchable() {
+        gpus.non_default()
+    } else {
+        gpus.get_default()
+    };
+
+    if let Some(gpu) = gpu {
+        for (opt, value) in gpu.launch_options() {
+            cmd.env(opt, value);
+        }
+    }
+
+    cmd
+}
+
+// either a command line argument or a field-code as described
+// in https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables
+enum ArgOrFieldCode<'a> {
+    SingleFileName,
+    FileList,
+    SingleUrl,
+    UrlList,
+    IconKey,
+    TranslatedName,
+    DesktopFileLocation,
+    Arg(&'a str),
+}
+
+impl<'a> TryFrom<&'a str> for ArgOrFieldCode<'a> {
+    type Error = ExecError<'a>;
+
+    fn try_from(value: &'a str) -> Result<Self, Self::Error> {
+        match value {
+            "%f" => Ok(ArgOrFieldCode::SingleFileName),
+            "%F" => Ok(ArgOrFieldCode::FileList),
+            "%u" => Ok(ArgOrFieldCode::SingleUrl),
+            "%U" => Ok(ArgOrFieldCode::UrlList),
+            "%i" => Ok(ArgOrFieldCode::IconKey),
+            "%c" => Ok(ArgOrFieldCode::TranslatedName),
+            "%k" => Ok(ArgOrFieldCode::DesktopFileLocation),
+            "%d" | "%D" | "%n" | "%N" | "%v" | "%m" => {
+                Err(ExecError::DeprecatedFieldCode(value.to_string()))
+            }
+            other if other.starts_with('%') => Err(ExecError::UnknownFieldCode(other.to_string())),
+            other => Ok(ArgOrFieldCode::Arg(other)),
+        }
+    }
+}
+
+// Returns the default terminal emulator linked to `/usr/bin/x-terminal-emulator`
+// or fallback to gnome terminal
+fn detect_terminal() -> (PathBuf, &'static str) {
+    use std::fs::read_link;
+
+    const SYMLINK: &str = "/usr/bin/x-terminal-emulator";
+
+    if let Ok(found) = read_link(SYMLINK) {
+        let arg = if found.to_string_lossy().contains("gnome-terminal") {
+            "--"
+        } else {
+            "-e"
+        };
+
+        return (read_link(&found).unwrap_or(found), arg);
+    }
+
+    (PathBuf::from("/usr/bin/gnome-terminal"), "--")
```

Comment:
Perhaps we should also check if `konsole` exists for those on KDE. Some GNOME installations now also won't have `gnome-terminal` anymore since they're replacing it with the libadwaita `gnome-console`.